File: Golang/Experimenter.go
package main

import (
	"fmt"
	"time"
	"sort"
	"reflect"
)

func RunExperirment(p Problem) Results{
	name := reflect.TypeOf(p).Name()
	fmt.Println("Running experiment on: "+name)
	measurements := []int64{}
	res := false;
	for i := 0; i<100000; i++ {
		now := time.Now()
		res = p.Calculate()
		end := time.Since(now)
		measurements = append(measurements, end.Nanoseconds())
		if !res {
			fmt.Println("Uhh algorithm failed")
		}
	}

	//PrintAll(measurements)
	return Results{avg:Average(measurements), med:Median(measurements), problemname:name, measurements:measurements, high:High(measurements), low:Low(measurements)}

}

func Median(durs []int64)int64{
	sort.Slice(durs, func(i, j int) bool {
		return durs[i] < durs[j]
	})
	return durs[len(durs)/2]
}

func High(durs []int64)int64{
	sort.Slice(durs, func(i, j int) bool {
		return durs[i] < durs[j]
	})
	return durs[len(durs)-1]
}

func Low(durs []int64)int64{
	sort.Slice(durs, func(i, j int) bool {
		return durs[i] < durs[j]
	})
	return durs[0]
}

func Average(durs []int64) int64{
	var total int64 = 0
	for _, x := range durs{
		total += x
	}
	return total/int64(len(durs))
}

/*
func PrintAll(durs []int64){
	for _,x := range durs{
		fmt.Print(x)
		fmt.Print(" - ")
	}
}
*/

File: Kotlin/src/Experimenter.kt
import kotlin.system.measureNanoTime

class Experimenter {

    fun RunExperiment(problem: Problem): ExperimentResult{

        println("Running experiment on: "+ problem.javaClass.name)

        var measurements = mutableListOf<Long>()

        for (i in 1..100000){
            var value = false
            var time = measureNanoTime {
                value = problem.Calculate()
            }
            measurements.add(time)
            if (!value){
                println("Algorithm failed")
            }
        }


        return ExperimentResult(problem.javaClass.name, measurements.toLongArray(), Avg(measurements.toLongArray()), Med(measurements.toLongArray()), High(measurements.toLongArray()), Low(measurements.toLongArray()));
    }


    private fun Avg(numbers: LongArray): Double {
        /*
        var total = 0L
        for (number in numbers){
            total += number
        }
        return total/numbers.size
        */

        return numbers.average()
    }

    private fun Med(numbers: LongArray): Long {
        numbers.sort()
        return numbers[numbers.size/2]
    }

    private fun High(numbers: LongArray): Long {
        numbers.sort()
        return numbers[numbers.size-1]
    }

    private fun Low(numbers: LongArray): Long {
        numbers.sort()
        return numbers[0]
    }


    fun AverageOfAverage(results: Array<ExperimentResult>):Double{

        var total: Double = 0.0
        for (result in results){
            total += result.avg
        }
        return total / results.size
    }
    fun AverageOfMedian(results: Array<ExperimentResult>):Double{

        var total: Double = 0.0
        for (result in results){
            total += result.med
        }
        return total / results.size
    }
    fun AverageOfHigh(results: Array<ExperimentResult>):Double{

        var total: Double = 0.0
        for (result in results){
            total += result.h
        }
        return total / results.size
    }
    fun AverageOfLow(results: Array<ExperimentResult>):Double{

        var total: Double = 0.0
        for (result in results){
            total += result.l
        }
        return total / results.size
    }

    fun PrintAverages(results: Array<ExperimentResult>){

        println("Printint Averages on: "+ results[0].Problem)
        println("Average-Average: "+Math.round(AverageOfAverage(results))+"ns")
        println("Average-Median: "+Math.round(AverageOfMedian(results))+"ns")
        println("Average-High: "+Math.round(AverageOfHigh(results))+"ns")
        println("Average-Low: "+Math.round(AverageOfLow(results))+"ns")


    }

    fun PrintIndi(results: Array<ExperimentResult>){
        for (result in results){
            result.PrintMe()
        }
    }



}